internal/plugintest/terraform_json_buffer.go

```diff
@@ -0,0 +1,125 @@
+package plugintest
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"io"
+	"log"
+	"regexp"
+
+	tfjson "github.com/hashicorp/terraform-json"
+)
+
+type TerraformJSONDiagnostics []tfjson.Diagnostic
+
+func (d TerraformJSONDiagnostics) Contains(r *regexp.Regexp, severity tfjson.DiagnosticSeverity) bool {
+	for _, v := range d {
+		if v.Severity != severity {
+			continue
+		}
+
+		if !r.MatchString(v.Summary) && !r.MatchString(v.Detail) {
+			continue
+		}
+
+		return true
+	}
+
+	return false
+}
+
+var _ io.Writer = &TerraformJSONBuffer{}
+var _ io.Reader = &TerraformJSONBuffer{}
+
+// TerraformJSONBuffer is used for storing and processing streaming
+// JSON output generated when running terraform commands with the
+// `-json` flag.
+type TerraformJSONBuffer struct {
+	buf         *bytes.Buffer
+	diagnostics TerraformJSONDiagnostics
+	jsonOutput  []string
+	rawOutput   string
```

Comment:
`rawOutput` was added in order to capture any outputs that were not JSON in order to maximise the information being returned to the provider developer for the purposes of debugging.

For example:
```shell

{"@level":"info","@message":"Terraform 1.3.2","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.322845Z","terraform":"1.3.2","type":"version","ui":"1.0"}
{"@level":"warn","@message":"Warning: Empty or non-existent state","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.355523Z","diagnostic":{"severity":"warning","summary":"Empty or non-existent state","detail":"There are currently no remote objects tracked in the state, so there is nothing to refresh."},"type":"diagnostic"}
{"@level":"info","@message":"Outputs: 0","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.355853Z","outputs":{},"type":"outputs"}
{"@level":"info","@message":"Terraform 1.3.2","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.380962Z","terraform":"1.3.2","type":"version","ui":"1.0"}
{"@level":"warn","@message":"Warning: Provider development overrides are in effect","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.381655Z","diagnostic":{"severity":"warning","summary":"Provider development overrides are in effect","detail":"The following provider development overrides are set in the CLI configuration:\n - bendbennett/playground in /Users/bdb/go/bin\n - bendbennett/timeouts in /Users/bdb/go/bin\n\nThe behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases."},"type":"diagnostic"}
{"@level":"info","@message":"example_resource.test: Plan to create","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.396541Z","change":{"resource":{"addr":"example_resource.test","module":"","resource":"example_resource.test","implied_provider":"example","resource_type":"example_resource","resource_name":"test","resource_key":null},"action":"create"},"type":"planned_change"}
{"@level":"info","@message":"Plan: 1 to add, 0 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.396576Z","changes":{"add":1,"change":0,"remove":0,"operation":"plan"},"type":"change_summary"}

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # example_resource.test will be created
  + resource "example_resource" "test" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
{"@level":"info","@message":"Terraform 1.3.2","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.479418Z","terraform":"1.3.2","type":"version","ui":"1.0"}
{"@level":"warn","@message":"Warning: Provider development overrides are in effect","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.480270Z","diagnostic":{"severity":"warning","summary":"Provider development overrides are in effect","detail":"The following provider development overrides are set in the CLI configuration:\n - bendbennett/playground in /Users/bdb/go/bin\n - bendbennett/timeouts in /Users/bdb/go/bin\n\nThe behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases."},"type":"diagnostic"}
{"@level":"info","@message":"example_resource.test: Plan to create","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.485951Z","change":{"resource":{"addr":"example_resource.test","module":"","resource":"example_resource.test","implied_provider":"example","resource_type":"example_resource","resource_name":"test","resource_key":null},"action":"create"},"type":"planned_change"}
{"@level":"info","@message":"example_resource.test: Creating...","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.491701Z","hook":{"resource":{"addr":"example_resource.test","module":"","resource":"example_resource.test","implied_provider":"example","resource_type":"example_resource","resource_name":"test","resource_key":null},"action":"create"},"type":"apply_start"}
{"@level":"info","@message":"example_resource.test: Creation errored after 0s","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.492349Z","hook":{"resource":{"addr":"example_resource.test","module":"","resource":"example_resource.test","implied_provider":"example","resource_type":"example_resource","resource_name":"test","resource_key":null},"action":"create","elapsed_seconds":0},"type":"apply_errored"}
{"@level":"error","@message":"Error: error summary","@module":"terraform.ui","@timestamp":"2023-01-25T13:54:15.503665Z","diagnostic":{"severity":"error","summary":"error summary","detail":"error detail","address":"example_resource.test","range":{"filename":"terraform_plugin_test.tf","start":{"line":1,"column":36,"byte":35},"end":{"line":1,"column":37,"byte":36}},"snippet":{"context":"resource \"example_resource\" \"test\"","code":"resource \"example_resource\" \"test\" { }","start_line":1,"highlight_start_offset":35,"highlight_end_offset":36,"values":[]}},"type":"diagnostic"}
```

Reply:
Could you expand a little more on when this output might be shown to provider developers? At first glance I'd be worried about showing this level of detail unless its potentially in TRACE logging from the testing code. There are separate environment variables for outputting Terraform core's logging, although they can be problematic sometimes.

I don't think it's necessarily a bad idea to capture this data though in case we do need it for advanced troubleshooting cases.